### PR TITLE
[Dialect/TensorRT] Update tensorrrt dialect to use non-dps calling convention

### PR DESCRIPTION
**Pull request by jhalakpatel**
Add non-DPS variant of `tensorrt::CallOp` i.e. `tensorrt::CallAllocOp`. Add conversion from call alloc op to runtime dialect enqueue alloc op.

### CODE REVIEW

File: mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp [hunk 1]
```diff
@@ -57,6 +57,80 @@ struct RewriteConstants : public OpRewritePattern<tensorrt::ConstantOp> {
   }
 };
 
+// Helper function to convert both CallOp and CallAllocOp
+static LogicalResult
+convertCallOp(Operation *op, MLIRContext *ctx, IRRewriter &rewriter,
```
Comment (christopherbate) [inline mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp:62]: Don't include `MLIRContext` in the argument list.... the context can be grabbed from any of the other arguments

File: mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp [hunk 2]
```diff
@@ -57,6 +57,80 @@ struct RewriteConstants : public OpRewritePattern<tensorrt::ConstantOp> {
...
+    llvm_unreachable("unexpected type of operation. Only callOp and "
+                     "callAllocOp are supported.");
+    return failure();
+  }
+
+  solver.load<dataflow::DeadCodeAnalysis>();
+  solver.load<dataflow::SparseConstantPropagation>();
+  solver.load<TensorKindAnalysis>(symbolTable);
+  if (failed(solver.initializeAndRun(trtFunc))) {
+    trtFunc.emitError() << "failed to run TensorKindAnalysis";
```
Comment (christopherbate) [inline mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp:89]: These two can be combined into one line `return trtFunc.emitError()...`

File: mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp [hunk 3]
```diff
@@ -57,6 +57,80 @@ struct RewriteConstants : public OpRewritePattern<tensorrt::ConstantOp> {
...
+            ? DenseI64ArrayAttr{}
+            : DenseI64ArrayAttr::get(ctx, hostTensorArgs));
+  } else {
+    enqueueOp = rewriter.create<trtrt::EnqueueAllocOp>(
+        loc, op->getResultTypes(), executionContext, stream, inputs,
+        /*host_tensors_args=*/hostTensorArgs.empty()
+            ? DenseI64ArrayAttr{}
+            : DenseI64ArrayAttr::get(ctx, hostTensorArgs));
+  }
+  rewriter.setInsertionPointAfter(enqueueOp);
```
Comment (christopherbate) [inline mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp:128]: Changing the insertion point here shouldn't have any effect. After creation of op, the insertion point is already after the op.

File: mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp [hunk 4]
```diff
@@ -72,6 +146,8 @@ class ConvertTensorRTToRuntimePass
     }
 
     IRRewriter rewriter(ctx);
+    SymbolTableCollection symbolTable;
+    DataFlowSolver solver;
```
Comment (christopherbate) [inline mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp:150]: You're doing something different there, keeping a single solver and repeatedly re-initializing it? I've never seen that pattern before, not sure if it will actually work as intended.

**Review by christopherbate — CHANGES_REQUESTED** [review]

File: mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp (hunk 4 above)
Comment (jhalakpatel) [reply mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp]: I did not realized that we need should not keep a single solver while iterating through different operations. I have reverted my change.

Comment (christopherbate) [reply mlir-tensorrt/compiler/lib/Conversion/TensorRTToTensorRTRuntime/TensorRTToTensorRTRuntime.cpp]: Yeah what needs to be done to keep a solver is to initialize it once. Then when you do replace/rewrite, you need to make corresponding small adjustments/remaps to the solver state. We do this in the Plan dialect transforms, and ideally we can do that here as well, but it's not especially high priority.

File: mlir-tensorrt/tensorrt/include/mlir-tensorrt-dialect/TensorRT/IR/TensorRTOps.td [hunk 5]
```diff
@@ -71,7 +71,9 @@ def TensorRT_TensorRTModuleOp : TensorRT_Op<"module", [
 
   let extraClassDeclaration = [{
     /// Return the name of this module if present.
-    std::optional<StringRef> getName() { return getSymName(); }
+    std::optional<StringRef> getName() const {
+      return (*this)->getAttrOfType<StringAttr>("sym_name").getValue();
```
Comment (christopherbate) [inline mlir-tensorrt/tensorrt/include/mlir-tensorrt-dialect/TensorRT/IR/TensorRTOps.td:75]: Why was this changed? Generally you don't want to rely on accessing properties through the string literal name.

File: mlir-tensorrt/tensorrt/lib/TensorRT/IR/TensorRT.cpp [hunk 6]
```diff
@@ -192,6 +192,39 @@ LogicalResult CallOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
...
+
+LogicalResult
+CallAllocOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  func::FuncOp kernel = getFuncCallee(symbolTable);
+  if (!kernel)
+    return emitOpError() << "no valid kernel found with symbol name "
+                         << getCallee();
+  FunctionType funcType = kernel.getFunctionType();
+
+  // TODO: For dynamic shapes, we allow passing a larger linear bufer.
```
Comment (christopherbate) [inline mlir-tensorrt/tensorrt/lib/TensorRT/IR/TensorRT.cpp:214]: Remove TODO? This TODO is talking about DPS, which we don't use here.

**Review by christopherbate — APPROVED** [review]
Minor comments otherwise LGTM

File: mlir-tensorrt/tensorrt/include/mlir-tensorrt-dialect/TensorRT/IR/TensorRTOps.td (hunk 5 above)
Comment (jhalakpatel) [reply mlir-tensorrt/tensorrt/include/mlir-tensorrt-dialect/TensorRT/IR/TensorRTOps.td:75]: It was required because we were passing trtModule why ArrayRef which I think is equivalent to passing by `const &`.
Accessing methods such as `getSymName` require the calling object to be `non const`.

I have reverted this change and instead now passing trtModule as `SmallVectorImpl<tensorrt::TensorRTModuleOp>& trtModules)`

**Comment by christopherbate** [comment]
@jhalakpatel Looks like it was merged without some of the changes you said you made. Fix-forward is fine, but current state is not good.
